### PR TITLE
repair: Allow ignoring dead nodes for replace operation

### LINKED ISSUE #9757: Allow ignoring dead nodes for replace operation
Consider

1) n1, n2, n3, n4, n5
2) n2 and n3 are both down
3) start n6 to replace n2

Replace operation in step 3 will fail because n3 is down.
We would see errors like below:

replace[25edeec0-57d4-11ec-be6b-7085c2409b2d]: Nodes={127.0.0.3} needed
for replace operation are down. It is highly recommended to fix the down
nodes and try again.

In the above example, currently, there is no way to replace any of the
dead nodes.

Users can either fix one of the dead nodes and run replace or run
removenode operation to remove one of the dead nodes.

Fixing dead nodes is always the best solution but it might not be
possible. Running removenode operation is not better than running
replace operation (with best effort by ignoring the other dead node) in
terms of data consistency. In addition, users have to run bootstrap
operation to add back the removed node. So, allowing replacing in such
case is a clear win.

### PR DESCRIPTION
**Pull request by asias**
Consider

1) n1, n2, n3, n4, n5
2) n2 and n3 are both down
3) start n6 to replace n2

Replace operation in step 3 will fail because n3 is down.
We would see errors like below:

replace[25edeec0-57d4-11ec-be6b-7085c2409b2d]: Nodes={127.0.0.3} needed
for replace operation are down. It is highly recommended to fix the down
nodes and try again.

In the above example, currently, there is no way to replace any of the
dead nodes.

Users can either fix one of the dead nodes and run replace or run
removenode operation to remove one of the dead nodes.

Fixing dead nodes is always the best solution but it might not be
possible. Running removenode operation is not better than running
replace operation (with best effort by ignoring the other dead node) in
terms of data consistency. In addition, users have to run bootstrap
operation to add back the removed node. So, allowing replacing in such
case is a clear win.

This patch adds the --ignore-dead-nodes-for-replace option to allow run
replace operation with best effort mode. Please note, use this option
only if the dead nodes are completely broken and down, and there is no
way to fix the node and bring it back. This also means the user has to
make sure the ignored dead nodes specified are really down to avoid any
data consistency issue.

Fixes #9757

### CODE REVIEW

File: service/storage_service.cc [hunk 1]
```diff
@@ -610,6 +610,26 @@ void storage_service::mark_existing_views_as_built() {
     }).get();
 }
 
+std::list<gms::inet_address> storage_service::get_ignore_dead_nodes_for_replace() {
+    std::vector<sstring> ignore_nodes_strs;
+    std::list<gms::inet_address> ignore_nodes;
+    boost::split(ignore_nodes_strs, _db.local().get_config().ignore_dead_nodes_for_replace(), boost::is_any_of(","));
```
Comment (bhalevy) [inline service/storage_service.cc:616]: you can strip spaces here too by including a space in `boost::is_any_of(",")` no?
that might create empty tokens but you skip them below anyhow...

**Review by bhalevy — APPROVED** [review]
Aside for a minor idea for improvement LGTM

**Comment by nyh** [comment]
@asias please rebase this PR as it has conflicts.

Also, I have a question. When a user decides to use `--ignore-dead-nodes-for-replace n3`, doesn't this mean that he lost hope for n3 to come back up? If this is the case, why not recommend that the user first removenode n3 and only then replacenode n2?

Or maybe the intended use case is that we do believe that n3 *will* come up eventually, and acknowledge that by ignoring it we will lose consistency (after n2's replace QUORUM reads may return old information) but are worried that if we wait any longer we risk losing a third node and losing actual data, not just QUORUM consistency?
If this is the explanation (or there is a different explanation?), we need to clearly document it where we explain the replacenode operation and what is the appropriate way to use it. Users will not be able to guess what is the tradeoff of using or not using the new "--ignore-dead-nodes-for-replace" option.

**Comment by asias** [comment]
> @asias please rebase this PR as it has conflicts.
> 
> Also, I have a question. When a user decides to use `--ignore-dead-nodes-for-replace n3`, doesn't this mean that he lost hope for n3 to come back up? If this is the case, why not recommend that the user first removenode n3 and only then replacenode n2?
> 
> Or maybe the intended use case is that we do believe that n3 _will_ come up eventually, and acknowledge that by ignoring it we will lose consistency (after n2's replace QUORUM reads may return old information) but are worried that if we wait any longer we risk losing a third node and losing actual data, not just QUORUM consistency? If this is the explanation (or there is a different explanation?), we need to clearly document it where we explain the replacenode operation and what is the appropriate way to use it. Users will not be able to guess what is the tradeoff of using or not using the new "--ignore-dead-nodes-for-replace" option.

Hello Nadav, the intended use case is to replace both of the dead nodes so that the cluster can return to the same state before the incident. It is the most common operation. Yes, when specifying nodes in the --ignore-dead-nodes-for-replace list, the user expect the nodes will not come back. 

In general, the preference is: fix the dead node if possible >  replace the dead node >  removenode the dead node

**Comment by asias** [comment]
Changes in v2:
- Rebased to latest master
- Expanded the comments a bit more

@nyh rebased.

**Comment by asias** [comment]
@nyh Ping. What else I can do?

File: service/storage_service.cc [hunk 2]
```diff
@@ -604,6 +604,26 @@ void storage_service::mark_existing_views_as_built() {
...
+    for (std::string n : ignore_nodes_strs) {
+        try {
+            std::replace(n.begin(), n.end(), '\"', ' ');
+            std::replace(n.begin(), n.end(), '\'', ' ');
```
Comment (nyh) [inline service/storage_service.cc:614]: Nitpick (don't have to fix now) I wish we didn't parse every list argument in a different function, and had a common function to parse all lists. How can a user guess that in this function list is supposed to be comma-separated, spaces are allowed around commas, and quotes are replaced by spaces - but in some other function there're different rules?

**Review by nyh — APPROVED** [review]
I'll approve and commit, because now I understand the use case, but I urge you to consider how you can provide better documentation for this later.

Only after reading all your replies to my questions did I understand that the purpose of this patch is to be able to use a "replace node" operation in the (unlikely but possible) case where two nodes died. This isn't actually written explicitly anywhere in comments or error messages below - I think the phrase "ignoring a node" is not helpful to understand this, because the user wouldn't think of this operation as "ignoring a node" but rather "replacing two dead nodes". I think comments and error messages could mention "replacing two dead nodes" to make things clearer.

Even more importantly, I think our user-facing documentation should mention this case of replacing two dead nodes, and how to do it with this new ignore_nodes option and what are the downsides of using it. You have scary messages about these downsides - but it's explained very shortly, and these messages don't give you a real explanation of the problem, or conversely - the non-problem: The thing is, if you **really** have two dead nodes which have no chance of coming back, you don't lose anything by ignore_nodes - you already lost consistency by losing two nodes, it's not the ignore_nodes which breaks the consistency. It just can't fix it.

**Comment by nyh** [comment]
> 

@vladzcloudius this might interest you, because you're one of the people who might want to know in the future how to replace **two** dead nodes, and I'm not sure how obvious it will be to you that you need to use this new "ignore dead nodes" option.

**Comment by vladzcloudius** [comment]
Thanks, @nyh.
It's more that obvious to me how relevant this is. This is super critical and it was not clear to me that we have an issue as described by @asias.

And I have some questions:
@asias I wonder why "removenode" would succeed in this case. 
Let's assume that you run "removenode" for n2 - what if n3 is one of the "receivers" of ranges from n2? Wouldn't it fail in that case? (*)

I thought that during "replacenode" scylla would try to find alive replicas for ranges, wouldn't it? (**)

If I'm right about (*) and wrong about (**) then this is a SUPER critical issue and it needs to be fixed ASAP because more than 1 node down at the same time is not so rare event.

**Pull request closed without merging**